src/tactic/linear_combination.lean

```diff
@@ -0,0 +1,392 @@
+
+
+/--
+This tactic changes the goal to be that the lefthand side of the target is
+  equal to the lefthand side of the given expression.  For example,
+  if `hsum_on_left` is `5*x - 5*y = 0`, and the target is `-5*y + 5*x = 0`, this
+  tactic will change the target to be `-5*y + 5*x = 5*x - 5*y`.
+
+This tactic only should be used when the target's type an equality whose
+  right side is 0.
+
+* Input:
+  * `hsum_on_left` : expr, whose type should be an equality with 0 on the right
+      side of the equals sign
+
+* Output: N/A
+-/
+meta def set_goal_to_hleft_eq_tleft (hsum_on_left : expr) : tactic unit :=
+do to_expr ``(replace_eq_expr %%hsum_on_left) >>= apply, skip
+
+/--
+This tactic attempts to prove the goal by normalizing the target if the
+`normalize` field of the given configuration is true.
+
+* Input:
+  * `config` : a `linear_combination_config`, which determines the tactic used
+      for normalization if normalization is done
+
+* Output: N/A
+-/
+meta def prove_equal_if_desired (config : linear_combination_config) :
+  tactic unit :=
+when config.normalize config.normalization_tactic
+
+/-! ### Part 4: Completed Tactic -/
+
+
+/--
+This is a tactic that attempts to prove the target by creating and applying a
+  linear combination of a list of equalities.  (If the `normalize` field of the
+  configuration is set to ff, then the tactic will simply set the user up to
+  prove their target using the linear combination instead of attempting to
+  finish the proof.)
+
+Note: The left and right sides of all the equalities should have the same
+  ring type, and the coefficients should also have this type.  There must be
+  instances of `has_mul` and `add_group` for this type.  Also note that the
+  target must involve at least one variable.
+
+* Input:
+  * `h_eqs_names` : a list of names, referring to equations in the local
+      context
+  * `coeffs` : a list of coefficients to be multiplied with the corresponding
+    equations in the list of names
+  * `config` : a `linear_combination_config`, which determines the tactic used
+      for normalization; by default, this value is the standard configuration
+      for a `linear_combination_config`
+
+* Output: N/A
+-/
+meta def linear_combination (h_eqs_names : list name) (coeffs : list pexpr)
+  (config : linear_combination_config := {}) : tactic unit :=
+do
+  hsum ← make_sum_of_hyps h_eqs_names coeffs,
+  hsum_on_left ← move_to_left_side hsum,
+  move_target_to_left_side,
+  set_goal_to_hleft_eq_tleft hsum_on_left,
+  prove_equal_if_desired config
+
+
+section interactive_mode
+setup_tactic_parser
+
+/--
+A parser that matches a pair in parentheses, where the first item in the pair
+is an identifier and the second item in the pair is a `pexpr`.
+
+* Input: None
+
+* Output: a `lean.parser (name × pexpr)`
+-/
+meta def parse_name_pexpr_pair : lean.parser (name × pexpr) :=
+do
+  tk "(",
+  id ← ident,
+  tk ",",
+  coeff ← parser.pexpr 0,
+  tk ")",
+  pure (id, coeff)
```

Comment:
A feature suggestion, if you think it makes sense: if an identifier appears by itself, then it gets interpreted as having coefficient 1.

I believe that all you need to implement this is to add an alternative case to this parsing routine:
```
meta def parse_name_pexpr_pair : lean.parser (name × pexpr) :=
(do
  tk "(",
  id ← ident,
  tk ",",
  coeff ← parser.pexpr 0,
  tk ")",
  pure (id, coeff)) <|>
(do
  id ← ident,
  pure (id, ``(1)))
```

Then, for example,
```
example (x y : ℤ) (h1 : x*y + 2*x = 1) (h2 : x = y) :
  x*y = -2*y + 1 :=
by linear_combination h1 (h2, -2)
```

Reply:
Thank you for this great suggestion!  I absolutely think this makes sense and would be a useful feature, so I just modified the code and added a few examples using this shortcut.